### PR TITLE
Set encoding explictly

### PR DESCRIPTION
**Pull request by crutchfield**
## 🗣 Description ##

Explicitly set encoding to UTF8 when Set-Content is invoked.

## 💭 Motivation and context ##

This standardizes encoding for files written to the file system. Primarily this is to ensure the OPA Agent is getting 
files encoded as UTF8. Closes issue [#20](https://github.com/cisagov/ScubaGear/issues/20)
  
## 🧪 Testing ##

Executed 'Invoke-SCuBA -ProductNames aad' in both PowerShell 5.1.19041.2364 and 7.3.0.  Verified ProviderSettingExport.json was utf8 encoded. 

## ✅ Pre-approval checklist ##
- [x ] This PR has an informative and human-readable title.
- [x ] Changes are limited to a single goal - *eschew scope creep!*
- [x ] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x ] All relevant type-of-change labels have been added.
- [x ] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x ] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x ] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x ] Tests have been added and/or modified to cover the changes in this PR.
- [x ] All new and existing tests pass.

## ✅ Pre-merge checklist ##
- [ X] Revert dependencies to default branches.
- [X ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Add a tag or create a release.

### CODE REVIEW

File: PowerShell/ScubaGear/Modules/Orchestrator.psm1 [hunk 1]
```diff
@@ -252,6 +252,24 @@ $ProdToFullName = @{
...
+    .Description
+    This function returns encoding type for setting content.
+    .Functionality
+    Internal
+    #>
+    $PSVersion = $PSVersionTable.PSVersion
+
+    $Encoding = 'utf8'
+
+    if ($PSVersion -ge '7.2.0'){
```
Comment (gdasher) [inline PowerShell/ScubaGear/Modules/Orchestrator.psm1:266]: It appears that powershell 6.0, 7.0 and 7.1 also default to (and supported) utf8NoBom.  Even though they are out of support, I suggest changing this check  emit utf8NoBom for those versions as well.

**Review by gdasher — COMMENTED** [review]
Thanks crutch.  Looks pretty good.  For your testing, did you test against a tenant that had any non-ASCII characters in it?  Also, did you confirm that rego handles the produced output correctly?

File: PowerShell/ScubaGear/Modules/Orchestrator.psm1 (hunk 1 above)
Comment (crutchfield) [reply PowerShell/ScubaGear/Modules/Orchestrator.psm1]: I updated to set encoding to utf8NoBom if PowerShell version is 6.0 or greater.

**Comment by gdasher** [comment]
The change looks good.  Were you able to reproduce the original issue? 

On my machine when I put a multibyte UTF8 character into the tenant (conditional access policy title) it got converted to a "?" with the code before your change, whereas the behavior in the bug was it properly making it into the rego but with an encoding that rego was unable to handle.  I have not yet tested with your code.

I will LGTM once I hear what testing you've been able to do.

**Comment by adhilto** [comment]
I made a conditional access policy with all the French special characters in the name, like so:
image

On main, the characters are replaced with "n++".
image


On this branch...it's something else:
image

Honestly, not sure which version is better. And ultimately, I wasn't able to replicate the error in the original issue either. Also curious to hear if you were able to recreate the original issue.

**Comment by crutchfield** [comment]
I have a unit test that used some non-ASCII characters.  This was executed against both PowerShell 5.1 and 7.3.  I also ran Invoke Scuba against AAD product and verified report was generated.  I did not change the tenant when running the test. 

 

Richard Crutchfield

Principal Software Systems Engineer

The MITRE Corporation

321-262-9246 (c)

407-737-9117 (w)

 

 <https://www.facebook.com/MITREcorp>  <https://www.linkedin.com/company/mitre>  <https://twitter.com/MITREcorp>  <https://www.youtube.com/user/mitrecorp>  <https://plus.google.com/+MitreOrgFFRDCs/posts> 

 <http://www.mitre.org/> 

 

 

From: Grant Dasher ***@***.***> 
Sent: Tuesday, January 24, 2023 6:02 PM
To: cisagov/ScubaGear ***@***.***>
Cc: Richard M Crutchfield ***@***.***>; Assign ***@***.***>
Subject: [EXT] Re: [cisagov/ScubaGear] Set encoding explictly (PR #131)

 

The change looks good. Were you able to reproduce the original issue?

On my machine when I put a multibyte UTF8 character into the tenant (conditional access policy title) it got converted to a "?" with the code before your change, whereas the behavior in the bug was it properly making it into the rego but with an encoding that rego was unable to handle. I have not yet tested with your code.

I will LGTM once I hear what testing you've been able to do.

**Comment by crutchfield** [comment]
I was not able to recreate the described issue.  I did attempt with several scenarios in a unit test but – no joy. ☹ 

 

Richard Crutchfield

Principal Software Systems Engineer

The MITRE Corporation

321-262-9246 (c)

407-737-9117 (w)

 

 <https://www.facebook.com/MITREcorp>  <https://www.linkedin.com/company/mitre>  <https://twitter.com/MITREcorp>  <https://www.youtube.com/user/mitrecorp>  <https://plus.google.com/+MitreOrgFFRDCs/posts> 

 <http://www.mitre.org/> 

 

 

From: Alden Hilton ***@***.***> 
Sent: Tuesday, January 24, 2023 11:15 PM
To: cisagov/ScubaGear ***@***.***>
Cc: Richard M Crutchfield ***@***.***>; Assign ***@***.***>
Subject: [EXT] Re: [cisagov/ScubaGear] Set encoding explictly (PR #131)

 

I made a conditional access policy with all the French special characters in the name, like so:
 <https://user-images.githubusercontent.com/106177711/214477784-55c1323b-cf12-45e2-98ce-5f985ae1e2bb.png> 

On main, the characters are replaced with "n++".
 <https://user-images.githubusercontent.com/106177711/214477883-1746c60e-4640-4ba0-9d2d-c1a644269fe3.png> 

On this branch...it's something else:
 <https://user-images.githubusercontent.com/106177711/214477941-5e73d7d7-a005-4c5a-8b24-e72e7291fb46.png> 

Honestly, not sure which version is better. And ultimately, I wasn't able to replicate the error in the original issue either. Also curious to hear if you were able to recreate the original issue.

**Review by gdasher — APPROVED** [review]
Ok, I don't think this is going to do any harm and hopefully it will fix the issue, even though we're unable to reproduce it.

Unit tests will be submitted after we create the structure in other cls?

**Comment by crutchfield** [comment]
I can add now or I will create an issue to add the unit test.  

 

I did add for the Configuration issue since we had talked.  They provided me with the file structures they were using so it should align. 

 

Richard Crutchfield

Principal Software Systems Engineer

The MITRE Corporation

321-262-9246 (c)

407-737-9117 (w)

 

 <https://www.facebook.com/MITREcorp>  <https://www.linkedin.com/company/mitre>  <https://twitter.com/MITREcorp>  <https://www.youtube.com/user/mitrecorp>  <https://plus.google.com/+MitreOrgFFRDCs/posts> 

 <http://www.mitre.org/> 

 

 

From: Grant Dasher ***@***.***> 
Sent: Wednesday, January 25, 2023 2:42 PM
To: cisagov/ScubaGear ***@***.***>
Cc: Richard M Crutchfield ***@***.***>; Assign ***@***.***>
Subject: [EXT] Re: [cisagov/ScubaGear] Set encoding explictly (PR #131)

 

@gdasher approved this pull request.

Ok, I don't think this is going to do any harm and hopefully it will fix the issue, even though we're unable to reproduce it.

Unit tests will be submitted after we create the structure in other commits?

**Review by adhilto — APPROVED** [review]

**Pull request merged**